### PR TITLE
BUG: DataFrame reductions losing EA dtypes

### LINKED ISSUE #42895: BUG: DataFrame[Int64].mean().dtype is object, should be Float64
This breaks the last corner case I want to test for #33036.

```
arr = np.random.randn(4, 3).astype("int64")
df = pd.DataFrame(arr).astype("Int64")
df.iloc[:, 1] = pd.NA  # <-- incorrectly casts to object, lets cast back and ignore that for now
df = df.astype("Int64")

res = df.mean()

>>> res
0     0.0
1    <NA>
2   -0.25
dtype: object
```

Trivial to fix with 2D EAs, not sure how/if to fix it without.

### LINKED ISSUE #51446: BUG: std with array_manager and NaT results
When a reduction results in NaT, the array manager reduction assumes the output dtype is the same as the input dtype. This is not correct for `std` (and from what I can tell, only `std` is incorrect).

```
df = pd.DataFrame({'a': ['2022-01-01', '2022-01-02', pd.NaT, '2022-01-03']})
df['a'] = pd.to_datetime(df['a'])

result = df.std(skipna=True)
print(result)
# a   1 days
# dtype: timedelta64[ns]

result2 = df.std(skipna=False)
print(result2)
# a   NaT
# dtype: datetime64[ns]
```

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #51446 (Replace xxxx with the GitHub issue number)
- [x] closes #42895
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

In the absence of 2D EAs, we need reductions to sometimes pretend the EA is 2D.  Enter "keepdims" adapted from numpy reductions.

This is still pretty ugly. I'm open to ideas to clean it up.

cc @rhshadrach any other particular cases need testing?

### CODE REVIEW

**Comment by jbrockmendel** [comment]
gentle ping @rhshadrach not looking to merge anytime soon but for thoughts on the approach

**Review by rhshadrach — COMMENTED** [review]
thanks for the ping - sorry this fell off my radar. I really like the approach here - It looks like this is close to impacting other ops like min, e.g.

```
df = pd.DataFrame({'a': [1, 1, 2], 'b': [3, 4, 5]}, dtype='Int64')
result = df.min()
print(result)  # currently is int64, should be Int64 I think
```

However, currently after making the values 2D, we do `values[~mask]` which brings it back to 1D again. I think we can instead do `np.where(mask, fill_value, values)`. This is similar to how nanops works.

> any other particular cases need testing?

Because of the above - it's not exactly clear what ops this is impacting. Could maybe add a basic test across the reduction ops and xfail any that lose EA dtype?

File: pandas/core/frame.py [hunk 1]
```diff
@@ -10906,13 +10906,25 @@ def func(values: np.ndarray):
...
+                        return vals2d._reduce(name, axis=1, skipna=skipna, **kwds)
+                    else:
+                        return values._reduce(name, axis=1, skipna=skipna, **kwds)
+
+                try:
+                    return values._reduce(name, skipna=skipna, keepdims=True, **kwds)
+                except (TypeError, ValueError):
+                    # no keepdims keyword yet; ValueError gets raised by
+                    #  util validator functions
+                    return values._reduce(name, skipna=skipna, **kwds)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/frame.py:10927]: right. there would need to be a deprecation cycle to allow 3rd party EAs to catch up

File: pandas/core/internals/array_manager.py [hunk 2]
```diff
@@ -976,24 +976,24 @@ def reduce(self, func: Callable) -> Self:
...
-            )
+            if isinstance(res, (np.ndarray, ExtensionArray)):
+                # keepdims worked!
+                result_arrays.append(res)
+            else:
+                # TODO NaT doesn't preserve dtype, so we need to ensure to create
+                # a timedelta result array if original was timedelta
+                # what if datetime results in timedelta? (eg std)
+                dtype = arr.dtype if res is NaT else None
+                result_arrays.append(sanitize_array([res], None, dtype=dtype))
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/internals/array_manager.py:991]: yes

**Comment by topper-123** [comment]
I've looked into this PR after you pointed it out. I had missed this PR unfortunately...
 
I like the keepdims on the arrays approach, it's a natural way to way to keep the dtype information that is being lost in reductions to scalars, rather than doing it on the frame as in my approach. I'm not a super fan of picking `"keepdims"` from the kwargs everywhere though, it seems brittle that it's necessary to add that to so many places.

I think adding a `keepdims=False` parameter to the signature of `ExtensionArray._reduce` and wrapping the returned scalar value there when `keepdims` is True before passing it on could be simpler? Then you wouldn't have to consider keepdims in all the reduction methods. That approach will also take care of wrapping the `NA` values that doesn't get wrapped currently.

Also, because I did #52707 from the point of solving #40669, I had a focus on frames with `min_count=1`. Many of the tests in #52707 fail in this PR currently (because NA result aren't wrapped in this current implementation), so those should be considered here also IMO.

I can take a closer look if the above approach works, if that's ok with you?

**Comment by jbrockmendel** [comment]
Closing in favor of #52788

**Pull request closed without merging**